File: apps/mobile-portfolio/ios/mobileportfolio/NativeSparkline/SparklinePoints.swift
import CoreGraphics

enum SparklinePoints {

    private static let pad: CGFloat = 0.06
    private static let amp: CGFloat = 0.35
    private static let placeholderCount = 10

    static func points(from values: [Double]) -> [CGPoint]? {
        guard values.count >= 2, let minV = values.min(), let maxV = values.max() else { return nil }
        let h = 1 - pad * 2
        let n = CGFloat(values.count - 1)
        let range = maxV - minV
        return values.enumerated().map { i, v in
            let x = CGFloat(i) / n
            let y: CGFloat = range > 0 ? 1 - pad - CGFloat((v - minV) / range) * h : 0.5
            return CGPoint(x: x, y: y)
        }
    }

    static func resolvedTrend(data: [Double]?, fallback: String) -> String {
        guard let arr = data, arr.count >= 2 else { return fallback }
        let first = arr.first!
        let last = arr.last!
        if last > first { return "up" }
        if last < first { return "down" }
        return "flat"
    }

    static func placeholderPoints(for trend: String, count: Int = placeholderCount) -> [CGPoint] {
        let n = max(count - 1, 1)
        var slope: CGFloat = 0
        if trend == "up" { slope = -1 }
        else if trend == "down" { slope = 1 }
        return (0..<count).map { i in
            let t = CGFloat(i) / CGFloat(n)
            let noise = sin(t * .pi * 2) * amp * 0.5
            let y = 0.5 + slope * t * amp + noise
            return CGPoint(x: t, y: y)
        }
    }
}

File: apps/mobile-portfolio/android/app/src/main/java/com/anonymous/mobileportfolio/view/sparkline/SparklinePoints.kt
package com.anonymous.mobileportfolio.view.sparkline

import kotlin.math.PI
import kotlin.math.sin

object SparklinePoints {

    private const val PAD = 0.06f
    private const val AMP = 0.35f
    private const val PLACEHOLDER_COUNT = 10

    fun scaleToPoints(values: DoubleArray): List<Pair<Float, Float>> {
        if (values.size < 2) return emptyList()
        val min = values.minOrNull() ?: 0.0
        val max = values.maxOrNull() ?: 1.0
        val range = max - min
        val height = 1 - PAD * 2
        val n = (values.size - 1).toFloat()
        val scale = if (range >= 1e-9) 1f / range.toFloat() else 0f
        return values.mapIndexed { i, v ->
            val x = i.toFloat() / n
            val y = if (scale > 0) 1 - PAD - ((v - min) * scale).toFloat() * height else 0.5f
            x to y
        }
    }

    fun resolveTrend(dataValues: DoubleArray?, fallbackTrend: String): String {
        if (dataValues == null || dataValues.size < 2) return fallbackTrend
        val first = dataValues.first()
        val last = dataValues.last()
        return when {
            last > first -> "up"
            last < first -> "down"
            else -> "flat"
        }
    }

    fun generatePlaceholder(trend: String, count: Int = PLACEHOLDER_COUNT): List<Pair<Float, Float>> {
        val n = (count - 1).coerceAtLeast(1).toFloat()
        val slope = when (trend) {
            "up" -> -1f
            "down" -> 1f
            else -> 0f
        }
        return (0 until count).map { i ->
            val t = i.toFloat() / n
            val noise = sin(t * PI.toFloat() * 2) * AMP * 0.5f
            val y = 0.5f + slope * t * AMP + noise
            t to y
        }
    }
}
